pkg/services/ngalert/notifier/channels/sensugo.go

```diff
@@ -0,0 +1,201 @@
+package channels
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"path"
+	"strings"
+	"time"
+
+	gokit_log "github.com/go-kit/kit/log"
+	"github.com/grafana/grafana/pkg/bus"
+	"github.com/grafana/grafana/pkg/infra/log"
+	"github.com/grafana/grafana/pkg/models"
+	"github.com/grafana/grafana/pkg/services/alerting"
+	old_notifiers "github.com/grafana/grafana/pkg/services/alerting/notifiers"
+	ngmodels "github.com/grafana/grafana/pkg/services/ngalert/models"
+	"github.com/prometheus/alertmanager/notify"
+	"github.com/prometheus/alertmanager/template"
+	"github.com/prometheus/alertmanager/types"
+	"github.com/prometheus/common/model"
+)
+
+type SensuGoNotifier struct {
+	old_notifiers.NotifierBase
+	log  log.Logger
+	tmpl *template.Template
+
+	URL       string
+	Entity    string
+	Check     string
+	Namespace string
+	Handler   string
+	APIKey    string
+	Message   string
+}
+
+// NewSlackNotifier is the constructor for the Slack notifier
+func NewSensuGoNotifier(model *models.AlertNotification, t *template.Template) (*SensuGoNotifier, error) {
+	if model.Settings == nil {
+		return nil, alerting.ValidationError{Reason: "No Settings Supplied"}
+	}
+
+	url := model.Settings.Get("url").MustString()
+	if url == "" {
+		return nil, alerting.ValidationError{Reason: "Could not find URL property in settings"}
+	}
+
+	apikey := model.DecryptedValue("apikey", model.Settings.Get("apikey").MustString())
+	if apikey == "" {
+		return nil, alerting.ValidationError{Reason: "Could not find the API Key property in settings"}
+	}
+
+	return &SensuGoNotifier{
+		NotifierBase: old_notifiers.NewNotifierBase(model),
+		URL:          url,
+		Entity:       model.Settings.Get("entity").MustString(),
+		Check:        model.Settings.Get("check").MustString(),
+		Namespace:    model.Settings.Get("namespace").MustString(),
+		Handler:      model.Settings.Get("handler").MustString(),
+		APIKey:       apikey,
+		Message:      model.Settings.Get("message").MustString(`{{ template "default.message" .}}`),
+		log:          log.New("alerting.notifier.slack"),
+		tmpl:         t,
+	}, nil
+}
+
+// Notify send alert notification to Sensu Go
+func (sn *SensuGoNotifier) Notify(ctx context.Context, as ...*types.Alert) (bool, error) {
+	sn.log.Info("Sending Sensu Go result")
+
+	data := notify.GetTemplateData(ctx, sn.tmpl, as, gokit_log.NewNopLogger())
+	var tmplErr error
+	tmpl := notify.TmplText(sn.tmpl, data, &tmplErr)
```

Comment:
There's no reason to not return immediately on error, right?
```suggestion
	var err error
	tmpl := notify.TmplText(sn.tmpl, data, &err)
	if err != nil {
		return false, err
	}
```

Reply:
The error is actually populated after we are done calling all the `tmpl(...)` functions below. So it should be checked after the last use of `tmpl(...)`.

Reply:
@codesome ah, I'll look again.

Reply:
@codesome yep you're right, I get how it works now.